icu4j/main/classes/core/src/com/ibm/icu/impl/number/MicroProps.java

```diff
@@ -38,17 +68,28 @@ public MicroProps(boolean immutable) {
         this.immutable = immutable;
     }
 
+    /**
+     * As MicroProps is the "base instance", this implementation of
+     * MircoPropsGenerator.processQuantity() just ensures that the output
+     * `micros` is correctly initialized.
+     * <p>
+     * For the "safe" invocation of this function, micros must not be *this,
+     * such that a copy of the base instance is made. For the "unsafe" path,
+     * this function can be used only once, because the base MicroProps instance
+     * will be modified and thus not be available for re-use.
+     *
+     * @param quantity The quantity for consideration and optional mutation.
+     * @return a MicroProps instance to populate.
+     */
     @Override
     public MicroProps processQuantity(DecimalQuantity quantity) {
         if (immutable) {
             return (MicroProps) this.clone();
-        } else if (exhausted) {
-            // Safety check
-            throw new AssertionError("Cannot re-use a mutable MicroProps in the quantity chain");
-        } else {
-            exhausted = true;
-            return this;
         }
+
+        assert !exhausted : "Cannot re-use a mutable MicroProps in the quantity chain";
+        exhausted = true;
+        return this;
```

Comment:
I feel uncomfortable with no longer throwing AssertionError in production code. I'm going to restore this safety check.

Reply:
Due to preferring caution, I've reinstated the old "throw AssertionError" via f01c11e53c570a3200ef1473e81abe45e5f573e0.